File: src/main/scala/matrixmult/matrixmult.scala
package sparsematrix

import java.io._
import java.util._
import scala.collection.mutable._
import scala.collection.immutable._
import scala.concurrent.forkjoin._
import scala.collection.parallel.immutable.ParRange
import scala.collection.parallel._

class DenseMatrix (_rows:Int, _columns:Int){
	val rows = _rows
	val cols = _columns
	val data = new Array[Int](rows*cols)

	def this(_rows:Int,_cols:Int,_in:BufferedReader) = {
		this(_rows,_cols)
		var i = 0
		while(i<rows){
			val line = _in.readLine
			val nums = line.split(" ").map(_.toInt)
			var j = 0
			while(j<cols){
				update(i,j,nums(j))
				j+=1
			}
			i+=1
		}
	}

	def update(_row : Int, _column : Int, _value : Int) = {
		data(_row*cols + _column) = _value
	}

	def apply(_row:Int,_column:Int):Int = {
		data(_row*cols + _column)
	}

	def print(out : PrintWriter) = {
		out.println(rows + " " + cols)
		var r = 0
		while(r<rows){
			val sb = new StringBuffer
			var c = 0
			while(c<cols){
//Why do these two lines give greatly different times?				
				sb.append(apply(r,c) + "\t")
//				sb.append("\t" + apply(r,c))
				c+=1
			}
			out.println(sb)
			r+=1
		}
	}
}

class SparseMatrix (val rows:Int, val cols:Int){
	val data = new Array[(Int,ArrayBuffer[(Int,Int)])](rows)
  var LOGGER = new PrintWriter(System.out)

	for(i <- 0 until rows)
		data(i)=(i,new ArrayBuffer[(Int,Int)])

	def this(_row:Int,_col:Int,_in:BufferedReader) = {
		this(_row,_col)
		var i=0
		while(i<rows){
			val (rowi,row) = data(i)
			val line = _in.readLine
      val nums :Array[Int] = line.split("\\s+").filter(_!="").map(_.toInt)
			var c = 0
			while(c<cols){
				if(nums(c)!=0)
					row+=((c,nums(c)))
				c+=1
			}
			i+=1
		}
	}

	def transpose : SparseMatrix = {
		val out = new SparseMatrix(cols,rows)
		var i = 0
		while(i<rows){
			val (_,row) = data(i)
			var j = 0
			while(j<row.length){
				val (c,v) = row(j)
				out(c,i) = v
				j+=1
			}
			i+=1
		}
		out
	}

	def update(_row:Int,_col:Int,_val:Int)= {
		data(_row)._2+=((_col,_val))
	}

	def apply(_row:Int,_col:Int):Int = {
		var i = 0
		while(i < data(_row)._2.length){
			val (c,v) = data(_row)._2(i)
			if(_col==c)
				return v
			i+=1
		}
		return 0
	}

	def print(out : PrintWriter) = {
		out.println(rows + " " + cols)
		var i = 0
		while(i<rows){
			val (_,row) = data(i)
			val sb = new StringBuffer
			var next = 0
			var j = 0
			while(j<row.length){
				val (c,v) = row(j)
				while(c>next){
					sb.append("0\t")
					sb.append(v + "\t")
					next+=1
				}
				j+=1
			}
			j = next
			while(j < cols){
				sb.append("0 ")
				j+=1
			}
			out.println(sb)
			i+=1
		}
	}

	def multiply(other:SparseMatrix,threads:Int=1):DenseMatrix = {
    LOGGER.println(System.currentTimeMillis + "\tStarting Parallel Multiplication with threads set to " + threads)
		val A = this
		if(A.rows!=other.cols)
			return null
    LOGGER.println(System.currentTimeMillis + "\tStarting Transpose")
		val B = other.transpose
    LOGGER.println(System.currentTimeMillis + "\tFinished Transpose")
		val C = new DenseMatrix(A.rows,B.rows)
		val range = threads match{
      case 0 => (0 until A.rows).par
      case 1 => (0 until A.rows)
      case n => {
        val r = (0 until A.rows).par
        r.tasksupport = new ForkJoinTaskSupport(new scala.concurrent.forkjoin.ForkJoinPool(threads))
        r
      }
    }

		for(i <- range){
			val (_,arow) = A.data(i)
			var j = 0
			while(j<B.rows){
				val (_,brow) = B.data(j)
				var sum = 0
				var aj = 0
				var bj = 0
				while(aj<arow.length&&bj<brow.length){
					if(arow(aj)._1==brow(bj)._1){
						sum += arow(aj)._2*brow(bj)._2
						aj+=1
						bj+=1
					}
					else if(arow(aj)._1<brow(bj)._1)
						aj+=1
					else
						bj+=1
				}
				C.data(i*C.cols+j)=sum
				j+=1
			}
		}
		return C
	}
}

File: src/main/java/matrixmult/MatrixMult.java
package matrixmult;

import java.util.Random;
import java.io.*;

import cern.colt.matrix.*;
import cern.colt.matrix.impl.*;

public class MatrixMult{

	public static void main(String[] args)throws IOException{
		if(args.length==0){
			System.out.println(
					"Java program for performing Sparse Matrix Multiplication"+
					"Usage: matrixmult.MatrixMult fileA fileB outputC [n_threads]\n"+
					"Where files are of the form\n"+
					"<n_rows> <n_columns> <n_entries>\n"+
					"followed on each line by the rows, with values space separated."
					);
			return;
		}
		testMineParallel(args);
	}

	/**
	 * Multiplies two sparse matrices using the (parallel?) COLT library from CERN.
	 *
	 * @param args arguments of the form &lt;input1&gt; &lt;input2&gt;
	 */
	public static void test(String[] args)throws IOException{
		System.out.println("SEQ");
		SparseDoubleMatrix2D mat = loadFileColt(args[0]);//new SparseDoubleMatrix2D(size,size);
		SparseDoubleMatrix2D mat2 =loadFileColt(args[1]);// new SparseDoubleMatrix2D(size,size);
		long start = System.currentTimeMillis();
		DoubleMatrix2D res = mat.zMult(mat2,null,1,1,false,false);
		long end = System.currentTimeMillis()-start;
		System.out.println("Time taken : " + end);
	}

	/**
	 * Multiplies two sparse matrices using the SparseMatrix class.
	 * Output is either written to a file (if a third argument is given) or sent to the screen.
	 *
	 * @param args arguments of the form &lt;input1&gt; &lt;input2&gt; &lt;outputFile&gt; [&lt;n_threads&gt;]
	 * @see matrixmult.SparseMatrix
	 */
	public static void testMineParallel(String[] args)throws IOException{
		try(PrintWriter logger = new PrintWriter(args[2]+".log")){
			logger.println(System.currentTimeMillis() + "\tStarting");

			SparseMatrix mat =  loadFileMM(args[0]);//new SparseMatrix(size,size);
			SparseMatrix mat2 =  loadFileMM(args[1]);//new SparseMatrix(size,size);
			SparseMatrix.LOGGER=logger;
			long start = System.currentTimeMillis();
			Matrix res;
			if(args.length==3){
				res = SparseMatrix.multiply(mat,mat2);
			}
			else{
				int threads = Integer.parseInt(args[3]);
				res = SparseMatrix.multiplyParallel(mat,mat2,threads);
			}
			long end = System.currentTimeMillis();
		
			logger.println(end + "\tMultiplying took " + (end-start));
			System.out.println("Time taken : " +(end-start));
			logger.println(System.currentTimeMillis() + "\tWriting output");
			res.print(new PrintWriter(args[2]));
			logger.println(System.currentTimeMillis() + "\tFinished");
		}
	}
	
	/**
	 * Loads a SparseMatrix from the given file according to the standard.
	 *
	 * @param fname The name of the file to load.
	 * @return The loaded matrix.
	 * @see matrixmult.SparseMatrix
	 */
	public static SparseMatrix loadFileMM(String fname)throws IOException{
		try(BufferedReader in = new BufferedReader(new FileReader(fname))){

			// Header should have three values, #rows, #columns, #entries
			String[] props = in.readLine().split(" ");
			int rows = Integer.parseInt(props[0]);
			int cols = Integer.parseInt(props[1]);
			int entries = Integer.parseInt(props[2]);

			SparseMatrix mat = new SparseMatrix(rows,cols);
			String line;
			int row =0;
			while((line=in.readLine())!=null){
				String[] nums = line.split("\\s+");
				int cnum = 0;
				for(int col=0;col<nums.length;col++){
					// White space and empty strings can sneak in. Ignore them.
					if(nums[col].equals("")||nums[col].equals(" "))
						continue;
					if(nums[col].equals("0")){
						cnum+=1;
						continue;
					}
					mat.set(row,cnum,Integer.parseInt(nums[col]));
					cnum+=1;
				}
				row++;
			}
			return mat;
		}
	}

	/**
	 * Loads a COLT sparse matrix from the given file according to the standard.
	 * @param fname The name of the file to load.
	 * @return The loaded matrix.
	 */
	public static SparseDoubleMatrix2D loadFileColt(String fname)throws IOException{
		try(BufferedReader in = new BufferedReader(new FileReader(fname))){
			
			// Header
			String[] props = in.readLine().split(" ");
			int rows = Integer.parseInt(props[0]);
			int cols = Integer.parseInt(props[1]);
			int entries = Integer.parseInt(props[2]);

			SparseDoubleMatrix2D mat = new SparseDoubleMatrix2D(rows,cols);
			String line;
			int row =0;
			while((line=in.readLine())!=null){
				String[] nums = line.split(" ");
				for(int col=0;col<nums.length;col++){
					if(!nums[col].equals("0")){
						mat.setQuick(row,col,Double.parseDouble(nums[col]));
					}
				}
				row++;
			}
			return mat;
		}
	}

}
